File: brainbox/longest_common_prefix.js
function longestCommonPrefix(strs) {
    if (!strs || strs.length === 0) return "";

    let prefix = strs[0];

    for (let i = 1; i < strs.length; i++) {
        let s = strs[i];

        while (!s.startsWith(prefix)) {
            prefix = prefix.slice(0, -1);
            if (prefix === "") return "";
        }
    }

    return prefix;
}


console.log(longestCommonPrefix(["flower","flow","flight"]));   // "fl"
console.log(longestCommonPrefix(["dog","racecar","car"]));      // ""
console.log(longestCommonPrefix(["interspecies","interstellar","interstate"]));  // "inters"
console.log(longestCommonPrefix(["throne","throne"]));          // "throne"
console.log(longestCommonPrefix(["throne","dungeon"]));         // ""
console.log(longestCommonPrefix(["a"]));                        // "a"
console.log(longestCommonPrefix(["",""]));                      // ""
console.log(longestCommonPrefix(["","abc"]));                   // ""
console.log(longestCommonPrefix(["abc",""]));                   // ""
console.log(longestCommonPrefix(["cir","car"]));                // "c"

File: brainbox/longest_common_prefix.py
def longestCommonPrefix(strs):
    if not strs:
        return ""

    prefix = strs[0]

    for s in strs[1:]:
        
        while not s.startswith(prefix):
            prefix = prefix[:-1]
            if prefix == "":
                return ""
    return prefix

print(longestCommonPrefix(["flower","flow","flight"]))   # "fl"
print(longestCommonPrefix(["dog","racecar","car"]))      # ""
print(longestCommonPrefix(["interspecies","interstellar","interstate"]))  # "inters"
print(longestCommonPrefix(["throne","throne"]))          # "throne"
print(longestCommonPrefix(["throne","dungeon"]))         # ""
print(longestCommonPrefix(["a"]))                        # "a"
print(longestCommonPrefix(["",""]))                      # ""
print(longestCommonPrefix(["","abc"]))                   # ""
print(longestCommonPrefix(["abc",""]))                   # ""
print(longestCommonPrefix(["cir","car"]))                # "c"
